Add tox 4 support

i386x opened the pull request:
Add a support for `tox` version 4, unit tests are still WIP.

i386x committed:
Rename legacy hooks to hooks3

tox 4 has a different API that implies rewriting almost the all
code base from scratch. The legacy code will be in hooks3 while
the tox 4 code will be in hooks4.

i386x committed:
Add missing mypy stubs

i386x committed:
Move generic code to standalone module

i386x committed:
Fix tests of legacy hooks

i386x committed:
Add unit tests for shared code

i386x committed:
Add implementation of hooks for tox 4

i386x committed:
Adjust project settings

- remove tox version restriction from install_requires
- exclude typing code and fallback imports from coverage reports
- exclude test_scripts directory from mypy
- add tox 4 to envlist
- add mypy to envlist
- do not run pydocstyle on typing stubs

richm commented:
I'm using fedora 37
```
>which tox
/usr/bin/tox
>rpm -qf /usr/bin/tox
tox-3.28.0-1.fc37.noarch
linux-system-roles/tox-lsr(tox-4)>python setup.py install --user --force
...
>pip list -v|grep tox-lsr
tox-lsr                  2.14.1         /home/rmeggins/.local/lib/python3.11/site-packages/tox_lsr-2.14.1-py3.11.egg
>tox -vv -r -e qemu-ansible-2.9 
using tox.ini: /home/rmeggins/linux-system-roles/mssql/tox.ini (pid 3038194)
  removing /home/rmeggins/linux-system-roles/mssql/.tox/log
using tox-3.28.0 from /usr/lib/python3.11/site-packages/tox/__init__.py (pid 3038194)
ERROR: No pyproject.toml or setup.py file found. The expected locations are:
  /home/rmeggins/linux-system-roles/mssql/pyproject.toml or /home/rmeggins/linux-system-roles/mssql/setup.py
You can
  1. Create one:
     https://tox.readthedocs.io/en/latest/example/package.html
  2. Configure tox to avoid running sdist:
     https://tox.readthedocs.io/en/latest/example/general.html
  3. Configure tox to use an isolated_build
```

i386x committed:
Add unit tests for tox 4 hooks

i386x committed:
Restrict python_version for tox 3

Removes pip install conflict.

i386x commented:
I gave it a try in a container:
```
[jkucera@fedora ~]$ podman run -it fedora:37
...
[root@3e0812c3caca /]# dnf -y install git python3-tox
...
[root@3e0812c3caca /]# cd ${HOME}
[root@3e0812c3caca ~]# git clone https://github.com/linux-system-roles/tox-lsr.git
...
[root@3e0812c3caca ~]# cd tox-lsr/
[root@3e0812c3caca tox-lsr]# git remote add i386x https://github.com/i386x/tox-lsr.git
[root@3e0812c3caca tox-lsr]# git fetch i386x
...
[root@3e0812c3caca tox-lsr]# git checkout tox-4
...
[root@3e0812c3caca tox-lsr]# pip install --user --force-reinstall -e .
Obtaining file:///root/tox-lsr
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting tox
  Downloading tox-4.4.12-py3-none-any.whl (148 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.0/149.0 kB 387.8 kB/s eta 0:00:00
Collecting configparser
  Downloading configparser-5.3.0-py3-none-any.whl (19 kB)
Collecting cachetools>=5.3
  Downloading cachetools-5.3.0-py3-none-any.whl (9.3 kB)
Collecting chardet>=5.1
  Downloading chardet-5.1.0-py3-none-any.whl (199 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.1/199.1 kB 1.2 MB/s eta 0:00:00
Collecting colorama>=0.4.6
  Downloading colorama-0.4.6-py2.py3-none-any.whl (25 kB)
Collecting filelock>=3.11
  Downloading filelock-3.11.0-py3-none-any.whl (10.0 kB)
Collecting packaging>=23
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 kB 2.5 MB/s eta 0:00:00
Collecting platformdirs>=3.2
  Downloading platformdirs-3.2.0-py3-none-any.whl (14 kB)
Collecting pluggy>=1
  Downloading pluggy-1.0.0-py2.py3-none-any.whl (13 kB)
Collecting pyproject-api>=1.5.1
  Downloading pyproject_api-1.5.1-py3-none-any.whl (12 kB)
Collecting virtualenv>=20.21
  Downloading virtualenv-20.21.0-py3-none-any.whl (8.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 489.7 kB/s eta 0:00:00
Collecting distlib<1,>=0.3.6
  Downloading distlib-0.3.6-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.5/468.5 kB 322.2 kB/s eta 0:00:00
Installing collected packages: distlib, pluggy, platformdirs, packaging, filelock, configparser, colorama, chardet, cachetools, virtualenv, pyproject-api, tox, tox-lsr
  Running setup.py develop for tox-lsr
Successfully installed cachetools-5.3.0 chardet-5.1.0 colorama-0.4.6 configparser-5.3.0 distlib-0.3.6 filelock-3.11.0 packaging-23.1 platformdirs-3.2.0 pluggy-1.0.0 pyproject-api-1.5.1 tox-4.4.12 tox-lsr-2.14.1 virtualenv-20.21.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

[root@3e0812c3caca tox-lsr]# tox --version
4.4.12 from /root/.local/lib/python3.11/site-packages/tox/__init__.py
registered plugins:
    tox-lsr-2.14.1 at /root/tox-lsr/src/tox_lsr/hooks.py

[root@3e0812c3caca tox-lsr]# cd ..
[root@3e0812c3caca ~]# git clone https://github.com/linux-system-roles/mssql.git
...
[root@3e0812c3caca ~]# cd mssql/
[root@3e0812c3caca mssql]# tox -vv -r -e qemu-ansible-2.9
ROOT: 98 D setup logging to DEBUG on pid 310 [tox/report.py:221]
qemu-ansible-2.9: 204 D created app data folder /root/.local/share/virtualenv [virtualenv/app_data/__init__.py:40]
qemu-ansible-2.9: 207 I find interpreter for spec PythonSpec(path=/usr/bin/python3) [virtualenv/discovery/builtin.py:56]
qemu-ansible-2.9: 207 I proposed PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:63]
qemu-ansible-2.9: 207 D accepted PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:65]
qemu-ansible-2.9: 208 D filesystem is case-sensitive [virtualenv/info.py:24]
qemu-ansible-2.9: 229 I create virtual environment via CPython3Posix(dest=/root/mssql/.tox/qemu-ansible-2.9, clear=False, no_vcs_ignore=False, global=False) [virtualenv/run/session.py:48]
qemu-ansible-2.9: 229 D create folder /root/mssql/.tox/qemu-ansible-2.9/bin [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 229 D create folder /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 229 D create folder /root/mssql/.tox/qemu-ansible-2.9/lib64/python3.11/site-packages [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 229 D write /root/mssql/.tox/qemu-ansible-2.9/pyvenv.cfg [virtualenv/create/pyenv_cfg.py:30]
qemu-ansible-2.9: 229 D 	home = /usr/bin [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 229 D 	implementation = CPython [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 229 D 	version_info = 3.11.2.final.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D 	virtualenv = 20.21.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D 	include-system-site-packages = false [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D 	base-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D 	base-exec-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D 	base-executable = /usr/bin/python3 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 230 D symlink /usr/bin/python3 to /root/mssql/.tox/qemu-ansible-2.9/bin/python [virtualenv/util/path/_sync.py:28]
qemu-ansible-2.9: 230 D create virtualenv import hook file /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_virtualenv.pth [virtualenv/create/via_global_ref/api.py:89]
qemu-ansible-2.9: 230 D create /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_virtualenv.py [virtualenv/create/via_global_ref/api.py:92]
qemu-ansible-2.9: 231 D ============================== target debug ============================== [virtualenv/run/session.py:50]
qemu-ansible-2.9: 231 D debug via /root/mssql/.tox/qemu-ansible-2.9/bin/python /root/.local/lib/python3.11/site-packages/virtualenv/create/debug.py [virtualenv/create/creator.py:193]
qemu-ansible-2.9: 231 D {
  "sys": {
    "executable": "/root/mssql/.tox/qemu-ansible-2.9/bin/python",
    "_base_executable": "/usr/bin/python3.11",
    "prefix": "/root/mssql/.tox/qemu-ansible-2.9",
    "base_prefix": "/usr",
    "real_prefix": null,
    "exec_prefix": "/root/mssql/.tox/qemu-ansible-2.9",
    "base_exec_prefix": "/usr",
    "path": [
      "/usr/lib64/python311.zip",
      "/usr/lib64/python3.11",
      "/usr/lib64/python3.11/lib-dynload",
      "/root/mssql/.tox/qemu-ansible-2.9/lib64/python3.11/site-packages",
      "/root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages"
    ],
    "meta_path": [
      "<class '_virtualenv._Finder'>",
      "<class '_frozen_importlib.BuiltinImporter'>",
      "<class '_frozen_importlib.FrozenImporter'>",
      "<class '_frozen_importlib_external.PathFinder'>"
    ],
    "fs_encoding": "utf-8",
    "io_encoding": "utf-8"
  },
  "version": "3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]",
  "makefile_filename": "/usr/lib64/python3.11/config-3.11-x86_64-linux-gnu/Makefile",
  "os": "<module 'os' (frozen)>",
  "site": "<module 'site' (frozen)>",
  "datetime": "<module 'datetime' from '/usr/lib64/python3.11/datetime.py'>",
  "math": "<module 'math' from '/usr/lib64/python3.11/lib-dynload/math.cpython-311-x86_64-linux-gnu.so'>",
  "json": "<module 'json' from '/usr/lib64/python3.11/json/__init__.py'>"
} [virtualenv/run/session.py:51]
qemu-ansible-2.9: 256 I add seed packages via FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/root/.local/share/virtualenv) [virtualenv/run/session.py:55]
qemu-ansible-2.9: 258 D install pip from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/pip-23.0.1-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 258 D install setuptools from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/setuptools-67.4.0-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 258 D install wheel from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/wheel-0.38.4-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 259 D build install image for setuptools-67.4.0-py3-none-any.whl to /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any [virtualenv/seed/embed/via_app_data/pip_install/base.py:45]
qemu-ansible-2.9: 259 D build install image for pip-23.0.1-py3-none-any.whl to /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any [virtualenv/seed/embed/via_app_data/pip_install/base.py:45]
qemu-ansible-2.9: 259 D build install image for wheel-0.38.4-py3-none-any.whl to /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any [virtualenv/seed/embed/via_app_data/pip_install/base.py:45]
qemu-ansible-2.9: 285 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel-0.38.4.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel-0.38.4.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 286 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 294 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel-0.38.4.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel-0.38.4.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 297 D generated console scripts wheel3.11 wheel wheel-3.11 wheel3 [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 342 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/distutils-precedence.pth to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/distutils-precedence.pth [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 342 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/pkg_resources to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pkg_resources [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 348 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/_distutils_hack to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_distutils_hack [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 349 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools-67.4.0.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools-67.4.0.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 349 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 396 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools-67.4.0.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools-67.4.0.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 398 D generated console scripts  [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 418 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 467 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip-23.0.1.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip-23.0.1.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 468 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip-23.0.1.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip-23.0.1.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 468 D generated console scripts pip pip3.11 pip-3.11 pip3 [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 468 I add activators for Bash, CShell, Fish, Nushell, PowerShell, Python [virtualenv/run/session.py:61]
qemu-ansible-2.9: 470 D write /root/mssql/.tox/qemu-ansible-2.9/pyvenv.cfg [virtualenv/create/pyenv_cfg.py:30]
qemu-ansible-2.9: 470 D 	home = /usr/bin [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	implementation = CPython [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	version_info = 3.11.2.final.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	virtualenv = 20.21.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	include-system-site-packages = false [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	base-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	base-exec-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 470 D 	base-executable = /usr/bin/python3 [virtualenv/create/pyenv_cfg.py:34]
  qemu-ansible-2.9: OK (0.28 seconds)
  congratulations :) (0.38 seconds)
[root@3e0812c3caca mssql]#
```
It looks it do something.

i386x commented:
It find the plugin but it did not load it. `--lsr-enable` is not listed in `tox --help`.

i386x committed:
Import marked hooks

Import functions marked with @impl or @hookimpl decorators so that
pluggy can see them.

i386x commented:
Now it seems that plugin works:
```
[root@3e0812c3caca mssql]# tox --help | grep lsr
           [--lsr-enable]
  --lsr-enable                Enable the use of the tox-lsr plugin (env: LSR_ENABLE) (default: None)

[root@3e0812c3caca mssql]# tox list
default environments:
black                       -> [no description]
pylint                      -> [no description]
flake8                      -> [no description]
yamllint                    -> [no description]
py26                        -> [no description]
py27                        -> [no description]
py36                        -> [no description]
py37                        -> [no description]
py38                        -> [no description]
py39                        -> [no description]
py310                       -> [no description]
py311                       -> [no description]
shellcheck                  -> [no description]
collection                  -> [no description]
ansible-lint                -> [no description]
custom                      -> [no description]
ansible-test                -> [no description]
woke                        -> [no description]
codeql                      -> [no description]

additional environments:
coveralls                   -> [no description]
coveralls26                 -> [no description]
molecule_version            -> [no description]
molecule_test               -> [no description]
molecule                    -> [no description]
qemu-ansible-2.9            -> [no description]
qemu-ansible-core-2.11      -> [no description]
qemu-ansible-core-2.12      -> [no description]
qemu-ansible-core-2.13      -> [no description]
qemu-ansible-core-2.14      -> [no description]
ansible-plugin-scan         -> [no description]
ansible-managed-var-comment -> [no description]
container-ansible-2.9       -> [no description]
container-ansible-core-2.12 -> [no description]
container-ansible-core-2.13 -> [no description]
container-ansible-core-2.14 -> [no description]
check-meta-versions         -> [no description]

[root@3e0812c3caca mssql]# tox -vv -r -e qemu-ansible-2.9
ROOT: 98 D setup logging to DEBUG on pid 519 [tox/report.py:221]
qemu-ansible-2.9: 207 W remove tox env folder /root/mssql/.tox/qemu-ansible-2.9 [tox/tox_env/api.py:322]
qemu-ansible-2.9: 524 I find interpreter for spec PythonSpec(major=3) [virtualenv/discovery/builtin.py:56]
qemu-ansible-2.9: 524 I proposed PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:63]
qemu-ansible-2.9: 524 D accepted PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:65]
qemu-ansible-2.9: 525 D filesystem is case-sensitive [virtualenv/info.py:24]
qemu-ansible-2.9: 546 I find interpreter for spec PythonSpec(major=3) [virtualenv/discovery/builtin.py:56]
qemu-ansible-2.9: 546 I proposed PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:63]
qemu-ansible-2.9: 546 D accepted PythonInfo(spec=CPython3.11.2.final.0-64, exe=/usr/bin/python3, platform=linux, version='3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]', encoding_fs_io=utf-8-utf-8) [virtualenv/discovery/builtin.py:65]
qemu-ansible-2.9: 547 I create virtual environment via CPython3Posix(dest=/root/mssql/.tox/qemu-ansible-2.9, clear=False, no_vcs_ignore=False, global=False) [virtualenv/run/session.py:48]
qemu-ansible-2.9: 547 D create folder /root/mssql/.tox/qemu-ansible-2.9/bin [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 547 D create folder /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 547 D create folder /root/mssql/.tox/qemu-ansible-2.9/lib64/python3.11/site-packages [virtualenv/util/path/_sync.py:9]
qemu-ansible-2.9: 548 D write /root/mssql/.tox/qemu-ansible-2.9/pyvenv.cfg [virtualenv/create/pyenv_cfg.py:30]
qemu-ansible-2.9: 548 D 	home = /usr/bin [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	implementation = CPython [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	version_info = 3.11.2.final.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	virtualenv = 20.21.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	include-system-site-packages = false [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	base-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	base-exec-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D 	base-executable = /usr/bin/python3 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 548 D symlink /usr/bin/python3 to /root/mssql/.tox/qemu-ansible-2.9/bin/python [virtualenv/util/path/_sync.py:28]
qemu-ansible-2.9: 548 D create virtualenv import hook file /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_virtualenv.pth [virtualenv/create/via_global_ref/api.py:89]
qemu-ansible-2.9: 549 D create /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_virtualenv.py [virtualenv/create/via_global_ref/api.py:92]
qemu-ansible-2.9: 549 D ============================== target debug ============================== [virtualenv/run/session.py:50]
qemu-ansible-2.9: 549 D debug via /root/mssql/.tox/qemu-ansible-2.9/bin/python /root/.local/lib/python3.11/site-packages/virtualenv/create/debug.py [virtualenv/create/creator.py:193]
qemu-ansible-2.9: 549 D {
  "sys": {
    "executable": "/root/mssql/.tox/qemu-ansible-2.9/bin/python",
    "_base_executable": "/usr/bin/python3.11",
    "prefix": "/root/mssql/.tox/qemu-ansible-2.9",
    "base_prefix": "/usr",
    "real_prefix": null,
    "exec_prefix": "/root/mssql/.tox/qemu-ansible-2.9",
    "base_exec_prefix": "/usr",
    "path": [
      "/usr/lib64/python311.zip",
      "/usr/lib64/python3.11",
      "/usr/lib64/python3.11/lib-dynload",
      "/root/mssql/.tox/qemu-ansible-2.9/lib64/python3.11/site-packages",
      "/root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages"
    ],
    "meta_path": [
      "<class '_virtualenv._Finder'>",
      "<class '_frozen_importlib.BuiltinImporter'>",
      "<class '_frozen_importlib.FrozenImporter'>",
      "<class '_frozen_importlib_external.PathFinder'>"
    ],
    "fs_encoding": "utf-8",
    "io_encoding": "utf-8"
  },
  "version": "3.11.2 (main, Feb  8 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)]",
  "makefile_filename": "/usr/lib64/python3.11/config-3.11-x86_64-linux-gnu/Makefile",
  "os": "<module 'os' (frozen)>",
  "site": "<module 'site' (frozen)>",
  "datetime": "<module 'datetime' from '/usr/lib64/python3.11/datetime.py'>",
  "math": "<module 'math' from '/usr/lib64/python3.11/lib-dynload/math.cpython-311-x86_64-linux-gnu.so'>",
  "json": "<module 'json' from '/usr/lib64/python3.11/json/__init__.py'>"
} [virtualenv/run/session.py:51]
qemu-ansible-2.9: 574 I add seed packages via FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/root/.local/share/virtualenv) [virtualenv/run/session.py:55]
qemu-ansible-2.9: 576 D got embed update of distribution pip from /root/.local/share/virtualenv/wheel/3.11/embed/3/pip.json [virtualenv/app_data/via_disk_folder.py:129]
qemu-ansible-2.9: 578 D got embed update of distribution setuptools from /root/.local/share/virtualenv/wheel/3.11/embed/3/setuptools.json [virtualenv/app_data/via_disk_folder.py:129]
qemu-ansible-2.9: 579 D got embed update of distribution wheel from /root/.local/share/virtualenv/wheel/3.11/embed/3/wheel.json [virtualenv/app_data/via_disk_folder.py:129]
qemu-ansible-2.9: 579 D install pip from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/pip-23.0.1-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 579 D install setuptools from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/setuptools-67.4.0-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 579 D install wheel from wheel /root/.local/lib/python3.11/site-packages/virtualenv/seed/wheels/embed/wheel-0.38.4-py3-none-any.whl via CopyPipInstall [virtualenv/seed/embed/via_app_data/via_app_data.py:47]
qemu-ansible-2.9: 581 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel-0.38.4.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel-0.38.4.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 581 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/distutils-precedence.pth to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/distutils-precedence.pth [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 581 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 581 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 582 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/pkg_resources to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pkg_resources [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 588 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/wheel-0.38.4-py3-none-any/wheel-0.38.4.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/wheel-0.38.4.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 591 D generated console scripts wheel3 wheel3.11 wheel wheel-3.11 [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 596 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/_distutils_hack to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/_distutils_hack [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 597 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools-67.4.0.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools-67.4.0.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 597 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 628 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/setuptools-67.4.0-py3-none-any/setuptools-67.4.0.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/setuptools-67.4.0.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 630 D generated console scripts  [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 656 D copy directory /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip-23.0.1.dist-info to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip-23.0.1.dist-info [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 657 D copy /root/.local/share/virtualenv/wheel/3.11/image/1/CopyPipInstall/pip-23.0.1-py3-none-any/pip-23.0.1.virtualenv to /root/mssql/.tox/qemu-ansible-2.9/lib/python3.11/site-packages/pip-23.0.1.virtualenv [virtualenv/util/path/_sync.py:36]
qemu-ansible-2.9: 658 D generated console scripts pip-3.11 pip pip3 pip3.11 [virtualenv/seed/embed/via_app_data/pip_install/base.py:41]
qemu-ansible-2.9: 658 I add activators for Bash, CShell, Fish, Nushell, PowerShell, Python [virtualenv/run/session.py:61]
qemu-ansible-2.9: 660 D write /root/mssql/.tox/qemu-ansible-2.9/pyvenv.cfg [virtualenv/create/pyenv_cfg.py:30]
qemu-ansible-2.9: 660 D 	home = /usr/bin [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	implementation = CPython [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	version_info = 3.11.2.final.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	virtualenv = 20.21.0 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	include-system-site-packages = false [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	base-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	base-exec-prefix = /usr [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 660 D 	base-executable = /usr/bin/python3 [virtualenv/create/pyenv_cfg.py:34]
qemu-ansible-2.9: 663 W install_deps> python -I -m pip install 'ansible==2.9.*' beautifulsoup4 'jinja2==2.7.*; python_version <= "3.7"' productmd PyYAML ruamel.yaml [tox/tox_env/api.py:428]
Ignoring jinja2: markers 'python_version <= "3.7"' don't match your environment
Collecting ansible==2.9.*
  Using cached ansible-2.9.27-py3-none-any.whl
Collecting beautifulsoup4
  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)
Collecting productmd
  Using cached productmd-1.35-py3-none-any.whl (45 kB)
Collecting PyYAML
  Using cached PyYAML-6.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (757 kB)
Collecting ruamel.yaml
  Using cached ruamel.yaml-0.17.21-py3-none-any.whl (109 kB)
Collecting jinja2
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
Collecting cryptography
  Using cached cryptography-40.0.2-cp36-abi3-manylinux_2_28_x86_64.whl (3.7 MB)
Collecting soupsieve>1.2
  Using cached soupsieve-2.4-py3-none-any.whl (37 kB)
Collecting six
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting cffi>=1.12
  Using cached cffi-1.15.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (462 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.1.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (27 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: soupsieve, six, ruamel.yaml, PyYAML, pycparser, MarkupSafe, productmd, jinja2, cffi, beautifulsoup4, cryptography, ansible
Successfully installed MarkupSafe-2.1.2 PyYAML-6.0 ansible-2.9.27 beautifulsoup4-4.12.2 cffi-1.15.1 cryptography-40.0.2 jinja2-3.1.2 productmd-1.35 pycparser-2.21 ruamel.yaml-0.17.21 six-1.16.0 soupsieve-2.4
qemu-ansible-2.9: 10354 I exit 0 (9.69 seconds) /root/mssql> python -I -m pip install 'ansible==2.9.*' beautifulsoup4 'jinja2==2.7.*; python_version <= "3.7"' productmd PyYAML ruamel.yaml pid=532 [tox/execute/api.py:275]
qemu-ansible-2.9: 10355 W commands[0]> python /root/tox-lsr/src/tox_lsr/test_scripts/runqemu.py [tox/tox_env/api.py:428]
CRITICAL:root:One, and only one, of --image-name or --image-file must be given.
qemu-ansible-2.9: 10491 C exit 1 (0.14 seconds) /root/mssql> python /root/tox-lsr/src/tox_lsr/test_scripts/runqemu.py pid=566 [tox/execute/api.py:275]
  qemu-ansible-2.9: FAIL code 1 (10.28=setup[10.15]+cmd[0.14] seconds)
  evaluation failed :( (10.39 seconds)
[root@3e0812c3caca mssql]#
```

richm commented:
yes - looks good so far - tox3 and tox4 are working on my laptop

spetrosi reviewed (approved):
This works on my Fedora 37

richm commented:
@i386x just need to figure out how to make pylint pass, then I think we can merge

i386x committed:
Run linters on ubuntu-latest

Run python linters only on ubuntu-latest due to differences in
results newer and later versions of linters are providing.

i386x commented:
@richm I adjusted `tox.yml` to run Python linters only on `ubuntu-latest` so the most recent versions of linters are used. All jobs are passing now, except the `python-26` one where the requested version of Ubuntu is unavailable.

richm commented:
ok - lgtm - @spetrosi please take a look
